torch/distributed/device_mesh.py

```diff
@@ -159,19 +159,41 @@ def create_sub_mesh(
 
             return res_submesh
 
-        def create_flatten_mesh(self, device_mesh: "DeviceMesh") -> "DeviceMesh":
+        def create_flatten_mesh(
+            self, device_mesh: "DeviceMesh", mesh_dim_name: Optional[str] = None
+        ) -> "DeviceMesh":
             root_mesh = _mesh_resources.get_root_mesh(device_mesh)
+
             flatten_dims_in_root = [
                 not_none(root_mesh.mesh_dim_names).index(flattened_mesh_dim_name)
                 for flattened_mesh_dim_name in not_none(device_mesh.mesh_dim_names)
             ]
-            flatten_mesh_dim_name = "_".join(
-                [
-                    not_none(root_mesh.mesh_dim_names)[dim]
-                    for dim in flatten_dims_in_root
-                ]
-            )
+
+            # Check whether the mesh_dim_name for flattened mesh is valid.
+            if mesh_dim_name:
+                self.flatten_name_to_root_dims.setdefault(root_mesh, {})
+                invalid_dim_names = chain(
+                    *list(not_none(root_mesh.mesh_dim_names)),
+                    *self.flatten_name_to_root_dims[root_mesh].keys(),
+                )
+                if mesh_dim_name in invalid_dim_names:
+                    raise RuntimeError(
+                        f"{mesh_dim_name} already exists for submesh of the {root_mesh}. ",
+                        f"The mesh_dim_names of submesh and flattened mesh are {invalid_dim_names}. "
+                        f"Please specify another valid mesh_dim_name.",
+                    )
+                else:
+                    flatten_mesh_dim_name = mesh_dim_name
+            else:
+                flatten_mesh_dim_name = "_".join(
+                    [
+                        not_none(root_mesh.mesh_dim_names)[dim]
+                        for dim in flatten_dims_in_root
+                    ]
+                )
```

Comment:
Even this auto flattening name may conflict with the existing names, right? Should we just unify the check?

Reply:
Hmm. I think auto flattening will never conflict with the existing names, right? Because we are concatenating the existing names.

Reply:
what if users create ["dp", "cp", "dp_cp"], and then flatten "dp" and "cp", lol. I know this sound unlikely to happen. But for completeness purpose, I don't see why we cannot verify both case.

Reply:
> what if users create ["dp", "cp", "dp_cp"], and then flatten "dp" and "cp", lol. I know this sound unlikely to happen. But for completeness purpose, I don't see why we cannot verify both case.

Oh you are right. I completely missed this case. lol